src/main/host/syscall/handler/clone.rs

```diff
@@ -64,55 +64,154 @@ impl SyscallHandler {
+            handled_flags.insert(CloneFlags::CLONE_SIGHAND);
+        }
+
+        if flags.contains(CloneFlags::CLONE_FS) {
+            // Currently a no-op since we don't support the related
+            // metadata and syscalls that this affects (e.g. chroot).
+            handled_flags.insert(CloneFlags::CLONE_FS);
+        }
+
+        if flags.contains(CloneFlags::CLONE_FILES) {
+            if !flags.contains(CloneFlags::CLONE_THREAD) {
+                // I *think* we'd just need to wrap the descriptor table
+                // in e.g. a RootedRc, and clone the Rc in this case.
+                warn!("Failing clone: we don't support fork with shared descriptor table");
+                return Err(Errno::ENOTSUP.into());
+            }
+            handled_flags.insert(CloneFlags::CLONE_FILES);
+        }
+
+        if flags.contains(CloneFlags::CLONE_SETTLS) {
+            native_flags.insert(CloneFlags::CLONE_SETTLS);
+            handled_flags.insert(CloneFlags::CLONE_SETTLS);
+        }
+
+        if flags.contains(CloneFlags::CLONE_VM) {
+            native_flags.insert(CloneFlags::CLONE_VM);
+            handled_flags.insert(CloneFlags::CLONE_VM);
+        }
+
+        if flags.contains(CloneFlags::CLONE_SYSVSEM) {
+            // Currently a no-op since we don't support sysv semaphores.
+            handled_flags.insert(CloneFlags::CLONE_SYSVSEM);
+        }
+
+        // Handled after native clone
+        let do_parent_settid = flags.contains(CloneFlags::CLONE_PARENT_SETTID);
+        handled_flags.insert(CloneFlags::CLONE_PARENT_SETTID);
+
+        // Handled after native clone
+        let do_child_settid = flags.contains(CloneFlags::CLONE_CHILD_SETTID);
+        handled_flags.insert(CloneFlags::CLONE_CHILD_SETTID);
+
+        // Handled after native clone
+        let do_child_cleartid = flags.contains(CloneFlags::CLONE_CHILD_CLEARTID);
+        handled_flags.insert(CloneFlags::CLONE_CHILD_CLEARTID);
+
+        let unhandled_flags = flags.difference(handled_flags);
+        if !unhandled_flags.is_empty() {
+            warn!("Unhandled clone flags: {unhandled_flags:?}");
+            return Err(Errno::ENOTSUP.into());
         }
 
         let child_tid = {
-            let flags_to_emulate = CloneFlags::CLONE_PARENT_SETTID
-                | CloneFlags::CLONE_CHILD_SETTID
-                | CloneFlags::CLONE_CHILD_CLEARTID;
-            let filtered_flags = flags.difference(flags_to_emulate);
             if raw_exit_signal != 0 {
                 warn!("Exit signal is unimplemented");
                 return Err(Errno::ENOTSUP.into());
             }
+
+            assert!(!native_flags
+                .contains(CloneFlags::CLONE_CHILD_SETTID | CloneFlags::CLONE_CHILD_CLEARTID));
+            let ctid = ForeignPtr::<libc::pid_t>::null();
+
+            assert!(!native_flags.contains(CloneFlags::CLONE_PARENT_SETTID));
+            let ptid = ForeignPtr::<libc::pid_t>::null();
```

Comment:
Add comment explaining what these assertions are for?

Reply:
Rearranged and added comments